src/dbnode/storage/coldflush.go

```diff
@@ -0,0 +1,198 @@
+		enabled:                true,
+	}
+}
+
+func (m *coldFlushManager) Disable() fileOpStatus {
+	m.Lock()
+	status := m.status
+	m.enabled = false
+	m.Unlock()
+	return status
+}
+
+func (m *coldFlushManager) Enable() fileOpStatus {
+	m.Lock()
+	status := m.status
+	m.enabled = true
+	m.Unlock()
+	return status
+}
+
+func (m *coldFlushManager) Status() fileOpStatus {
+	m.RLock()
+	status := m.status
+	m.RUnlock()
+	return status
+}
+
+func (m *coldFlushManager) Run(t time.Time) bool {
+	m.Lock()
+	if !m.shouldRunWithLock() {
+		m.Unlock()
+		return false
+	}
+	m.status = fileOpInProgress
+	m.Unlock()
+
+	// NB(xichen): perform data cleanup and flushing sequentially to minimize the impact of disk seeks.
+	// NB(r): Use invariant here since flush errors were introduced
+	// and not caught in CI or integration tests.
+	// When an invariant occurs in CI tests it panics so as to fail
+	// the build.
+	if err := m.ColdFlushCleanup(t, m.database.IsBootstrapped()); err != nil {
+		instrument.EmitAndLogInvariantViolation(m.opts.InstrumentOptions(),
+			func(l *zap.Logger) {
+				l.Error("error when cleaning up data", zap.Time("time", t), zap.Error(err))
+			})
+	}
+	if err := m.coldFlush(); err != nil {
+		instrument.EmitAndLogInvariantViolation(m.opts.InstrumentOptions(),
+			func(l *zap.Logger) {
+				l.Error("error when flushing data", zap.Time("time", t), zap.Error(err))
+			})
+	}
+	m.Lock()
+	m.status = fileOpNotStarted
+	m.Unlock()
+	return true
+}
+
+func (m *coldFlushManager) coldFlush() error {
+	namespaces, err := m.database.OwnedNamespaces()
+	if err != nil {
+		return err
+	}
+
+	// The cold flush process will persist any data that has been "loaded" into memory via
+	// the Load() API but has not yet been persisted durably. As a result, if the cold flush
+	// process completes without error, then we want to "decrement" the number of tracked bytes
+	// by however many were outstanding right before the cold flush began.
+	//
+	// For example:
+	// t0: Load 100 bytes --> (numLoadedBytes == 100, numPendingLoadedBytes == 0)
+	// t1: memTracker.MarkLoadedAsPending() --> (numLoadedBytes == 100, numPendingLoadedBytes == 100)
+	// t2: Load 200 bytes --> (numLoadedBytes == 300, numPendingLoadedBytes == 100)
+	// t3: ColdFlushStart()
+	// t4: Load 300 bytes --> (numLoadedBytes == 600, numPendingLoadedBytes == 100)
+	// t5: ColdFlushEnd()
+	// t6: memTracker.DecPendingLoadedBytes() --> (numLoadedBytes == 500, numPendingLoadedBytes == 0)
+	// t7: memTracker.MarkLoadedAsPending() --> (numLoadedBytes == 500, numPendingLoadedBytes == 500)
+	// t8: ColdFlushStart()
+	// t9: ColdFlushError()
+	// t10: memTracker.MarkLoadedAsPending() --> (numLoadedBytes == 500, numPendingLoadedBytes == 500)
+	// t11: ColdFlushStart()
+	// t12: ColdFlushEnd()
+	// t13: memTracker.DecPendingLoadedBytes() --> (numLoadedBytes == 0, numPendingLoadedBytes == 0)
+	memTracker := m.opts.MemoryTracker()
+	memTracker.MarkLoadedAsPending()
+
+	defer func() {
+		if err == nil {
```

Comment:
I know this was ported, but this err check looks kind of suss.

Perhaps we wrap an inner method to do the actual "coldFlush" and then the outer method "trackedColdFlush" or something does the whole "MarkLoadedAsPending()" and "if err := m.coldFlush(); err != nil { memTracker.DecPendingLoadedBytes() }"?  That's similar to how we do things elsewhere and doesn't accidentally stop working if the "err" variable is shadowed and not actually set to a non-nil value correctly within the scope that it's being read in this defer function.